manta-crypto/src/arkworks/serialize.rs

```diff
@@ -25,6 +28,153 @@ use {
+    type Serialize: CanonicalSerialize + From<&'s Self>;
+}
+
+/// Deserialization Hook
+pub trait HasDeserialization: Sized {
+    /// Deserialize Type
+    type Deserialize: CanonicalDeserialize + Into<Self>;
+}
+
+/// Arkworks Reader
+pub struct ArkReader<R>
+where
+    R: codec::Read,
+{
+    /// Reader State
+    state: Result<R, R::Error>,
+}
+
+impl<R> ArkReader<R>
+where
+    R: codec::Read,
+{
+    /// Builds a new [`ArkReader`] from `reader`.
+    #[inline]
+    pub fn new(reader: R) -> Self {
+        Self { state: Ok(reader) }
+    }
+
+    /// Updates the internal reader state by performing the `f` computation.
+    #[inline]
+    fn update<T, F>(&mut self, f: F) -> Option<T>
+    where
+        F: FnOnce(&mut R) -> Result<T, R::Error>,
+    {
+        if let Ok(reader) = self.state.as_mut() {
+            match f(reader) {
+                Ok(value) => return Some(value),
+                Err(err) => self.state = Err(err),
+            }
+        }
+        None
+    }
+
+    /// Returns the reader state back or an error if it occured during any [`Read`](io::Read)
+    /// methods.
+    #[inline]
+    pub fn finish(self) -> Result<R, R::Error> {
+        self.state
+    }
+}
+
+impl<R> io::Read for ArkReader<R>
+where
+    R: codec::Read,
+{
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        self.update(|reader| reader.read(buf))
+            .ok_or_else(|| Error::new(ErrorKind::Other, "Reading Error"))
+    }
+
+    #[inline]
+    fn read_exact(&mut self, buf: &mut [u8]) -> Result<(), Error> {
+        match self.update(|reader| match reader.read_exact(buf) {
+            Ok(value) => Ok(Ok(value)),
+            Err(ReadExactError::Read(err)) => Err(err),
+            Err(ReadExactError::UnexpectedEnd(err)) => Ok(Err(err)),
+        }) {
+            Some(Ok(_)) => Ok(()),
+            Some(Err(_)) => Err(Error::new(
+                ErrorKind::UnexpectedEof,
+                "Unexpected end of buffer.",
+            )),
+            _ => Err(Error::new(ErrorKind::Other, "Reading Error")),
+        }
+    }
+}
+
+/// Arkworks Writer
+pub struct ArkWriter<W>
+where
+    W: codec::Write,
+{
+    /// Writer State
+    state: Result<W, W::Error>,
+}
+
+impl<W> ArkWriter<W>
+where
+    W: codec::Write,
+{
+    /// Builds a new [`ArkWriter`] from `writer`.
+    #[inline]
+    pub fn new(writer: W) -> Self {
+        Self { state: Ok(writer) }
+    }
+
+    /// Updates the internal writer state by performing the `f` computation.
+    #[inline]
+    fn update<T, F>(&mut self, f: F) -> Option<T>
+    where
+        F: FnOnce(&mut W) -> Result<T, W::Error>,
+    {
+        if let Ok(writer) = self.state.as_mut() {
+            match f(writer) {
+                Ok(value) => return Some(value),
+                Err(err) => self.state = Err(err),
+            }
+        }
+        None
+    }
+
+    /// Returns the writer state back or an error if it occured during any [`Write`](io::Write)
```

Comment:
The state is defined as the result, not `W`.

Reply:
Yeah but that's an internal detail, it's a convenient way to carry around the error.